src/foam/nanos/ruler/Rule.js

```diff
@@ -230,6 +234,48 @@
           ruler.getDelegate().cmd_(x.put("OBJ", obj), getCmd());
         }
       `
+    },
+    {
+      name: 'reverseAction',
+      args: [
+        {
+          name: 'obj',
+          type: 'foam.core.FObject'
+        },
+        {
+          name: 'x',
+          type: 'foam.core.X'
+        }
+      ],
+      documentation: 'if one of the rules in a group throws an exception we need a way to reverse actions for previously executed rules.',
+      javaCode: ` `
+    },
+    {
+      name: 'updateRule',
+      type: 'foam.nanos.ruler.Rule',
+      documentation: 'since rules are stored as lists in the RulerDAO we use listeners to update them whenever ruleDAO is updated.' +
+      'the method provides logic for modifying already stored rule. If not overridden, the incoming rule will be stored in the list as it is.',
+      args: [
+        {
+          name: 'rule',
+          type: 'foam.nanos.ruler.Rule'
+        }
+      ],
+      javaCode: `
+      return rule;`
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(`
+        public static Rule findById(Collection<Rule> listRule, Long passedId) {
```

Comment:
It is going to be slow to do a list scan every time. Couldn't you store in a Map to improve lookup performance?

Reply:
Not an issue.  Only done when rule dao changes.